source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc

```diff
@@ -0,0 +1,179 @@
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateOutgoingUnidirectionalStream() {
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void EnvoyQuicServerSession::setUpRequestDecoder(EnvoyQuicStream& stream) {
+  ASSERT(http_connection_callbacks_ != nullptr);
+  Http::StreamDecoder& decoder = http_connection_callbacks_->newStream(stream);
+  stream.setDecoder(decoder);
+}
+
+void EnvoyQuicServerSession::OnConnectionClosed(const quic::QuicConnectionCloseFrame& frame,
+                                                quic::ConnectionCloseSource source) {
+  quic::QuicServerSessionBase::OnConnectionClosed(frame, source);
+  for (auto callback : network_connection_callbacks_) {
+    // Tell filters about connection close.
+    callback->onEvent(source == quic::ConnectionCloseSource::FROM_PEER
+                          ? Network::ConnectionEvent::RemoteClose
+                          : Network::ConnectionEvent::LocalClose);
+  }
+  transport_failure_reason_ = absl::StrCat(quic::QuicErrorCodeToString(frame.quic_error_code),
+                                           " with details: ", frame.error_details);
+}
+
+void EnvoyQuicServerSession::Initialize() {
+  quic::QuicServerSessionBase::Initialize();
+  reinterpret_cast<EnvoyQuicConnection*>(connection())->setEnvoyConnection(*this);
+}
+
+void EnvoyQuicServerSession::addWriteFilter(Network::WriteFilterSharedPtr filter) {
+  filter_manager_.addWriteFilter(filter);
+}
+
+void EnvoyQuicServerSession::addFilter(Network::FilterSharedPtr filter) {
+  filter_manager_.addFilter(filter);
+}
+
+void EnvoyQuicServerSession::addReadFilter(Network::ReadFilterSharedPtr filter) {
+  filter_manager_.addReadFilter(filter);
+}
+
+bool EnvoyQuicServerSession::initializeReadFilters() {
+  return filter_manager_.initializeReadFilters();
+}
+
+void EnvoyQuicServerSession::addConnectionCallbacks(Network::ConnectionCallbacks& cb) {
+  network_connection_callbacks_.push_back(&cb);
+}
+
+void EnvoyQuicServerSession::addBytesSentCallback(Network::Connection::BytesSentCb /*cb*/) {
+  // Proxy is not supported.
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+void EnvoyQuicServerSession::enableHalfClose(bool enabled) {
+  ASSERT(!enabled, "Quic connection doesn't support half close.");
+}
+
+void EnvoyQuicServerSession::setBufferLimits(uint32_t /*limit*/) {
+  // TODO(danzh): add interface to quic for connection level buffer throttling.
+  // Currently read buffer is capped by connection level flow control. And
+  // write buffer is not capped.
+  ENVOY_CONN_LOG(error, "Quic manages its own buffer currently.", *this);
+}
+
+uint32_t EnvoyQuicServerSession::bufferLimit() const {
+  // As quic connection is not HTTP1.1, this method shouldn't be called by HCM.
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void EnvoyQuicServerSession::close(Network::ConnectionCloseType type) {
+  // TODO(danzh): Implement FlushWrite and FlushWriteAndDelay mode.
+  ASSERT(type == Network::ConnectionCloseType::NoFlush);
```

Comment:
I suspect the assert won't work unless we tweak the HCM - maybe log an error and just close?

Reply:
I regretted to change HCM. Change to error log here instead.